src/data/markdown/docs/01 guides/05 Testing Guides/04 Large-scale k6 testing.md

```diff
@@ -0,0 +1,484 @@
+
+### EC2 costs
+The AWS EC2 instances are relatively cheap. Even the largest instance we have used in this benchmark (m5.24xlarge) costs only $4.6 per hour. Make sure to turn off the load-gen servers once you are done with your testing. Forgotten EC2 server will cost $3312 per month.  
+Tip: it's often possible to launch "spot instances" of the same hardware for 10-20% of the cost. 
+
+## Errors
+
+If you run into errors during the execution, it's good to understand if they were caused by the load-generator or by the failing SUT. 
+
+### read: connection reset by peer
+
+Error similar to this one is caused by the target system resetting the TCP connection. This happens when the Load balancer or the server itself isn't able to handle the traffic.
+
+<div class="code-group">
+
+```text
+WARN[0013] Request Failed       error="Get http://test.k6.io: read tcp 172.31.72.209:35288->63.32.205.136:80: read: connection reset by peer"
+```
+
+</div>
+
+### context deadline exceeded
+
+Error like this happens when k6 was able to send a request, but the target system didn't respond in time. The default timeout in k6 is 60 seconds. If your system doesn't produce the response in this timeframe, this error will appear.
+
+<div class="code-group">
+
+```
+WARN[0064] Request Failed    error="Get http://test.k6.io: context deadline exceeded"
+```
+
+</div>
+
+### dial tcp 52.18.24.222:80: i/o timeout
+
+This is a similar error to the one above, but in this case k6 wasn't even able to make a request. The target system isn't able to establish a connection.
+
+
+<div class="code-group">
+
+```
+WARN[0057] Request Failed     error="Get http://pawel.staging.loadimpact.com/static/logo.svg?url=v3: dial tcp 52.18.24.222:80: i/o timeout"
+```
+
+</div>
+
+### socket: too many open files
+
+This error means that the load-generator machine isn't able to open TCP sockets because it reached the limit of open file descriptors.
+Make sure that your limit is set sufficiently high `ulimit -n 250000` should be enough for anyone :tm:
+
+<div class="code-group">
+
+```
+WARN[0034] Request Failed     error="Get http://99.81.83.131/static/logo.svg?ip=6: dial tcp 99.81.83.131:80: socket: too many open files"
+```
+
+</div>
+
+
+Note: you should decide what level of errors is acceptable. At large scale some errors are always present.
+If you make 50M requests with 100 failures, this is generally a good result (0.00002% errors).
+
+# Benchmarking k6 on AWS hardware
+
+We have executed few large tests on different EC2 machines to see how much load k6 can generate. 
+Our general observation is that k6 scales proportionally to the hardware. 2x larger machine is able to generate 2x more traffic.
+The limit to this scalability is in the number of open connections. Single Linux machine can open up to `65 535` sockets per IP. 
```

Comment:
nitpick but it's per <src IP, src port, dst IP, dst port> tuple afaik, so should be able to go beyond 65k if machine has multiple nics with different IPs or target system hostnam is behind a load balancer with different IPs etc.

Reply:
I'm happy to update that, but I think I'll have to mention that k6 doesn't have support for multiple IPs and discuss this for a while. I don't want to go too much off-topic :-) (here's another hyphen!)

I also think that the limit of 65k is for <src IP, src port>. I don't think the destination IP matters when all 65k ports are already used by the load-gen server. (or maybe I'm missing something).